File: src/shared/transfer-crypto.ts
const textEncoder = new TextEncoder();
const kdfSaltPrefix = "vidyut-v1-file-transfer";
const kdfIterations = 200_000;
const keyCache = new Map<string, Promise<CryptoKey>>();

export interface TransferChunkMetadata {
  transferId: string;
  fileId: string;
  offset: number;
  plaintextBytes: number;
}

export interface EncryptedTransferChunk extends TransferChunkMetadata {
  nonce: string;
  ciphertext: Uint8Array;
}

export async function encryptTransferChunk(
  metadata: TransferChunkMetadata,
  plaintext: Uint8Array,
  pairingSecret: string,
  nonceBytes: Uint8Array = crypto.getRandomValues(new Uint8Array(12)),
): Promise<EncryptedTransferChunk> {
  if (plaintext.byteLength !== metadata.plaintextBytes) {
    throw new RangeError("Chunk byte count does not match its metadata.");
  }
  const nonce = toBase64(nonceBytes);
  const ciphertext = await crypto.subtle.encrypt(
    {
      name: "AES-GCM",
      iv: toArrayBuffer(nonceBytes),
      additionalData: associatedData(metadata, nonce),
    },
    await deriveTransferKey(pairingSecret, metadata.transferId),
    toArrayBuffer(plaintext),
  );
  return {
    ...metadata,
    nonce,
    ciphertext: new Uint8Array(ciphertext),
  };
}

export async function decryptTransferChunk(
  chunk: EncryptedTransferChunk,
  pairingSecret: string,
): Promise<Uint8Array> {
  const plaintext = new Uint8Array(
    await crypto.subtle.decrypt(
      {
        name: "AES-GCM",
        iv: toArrayBuffer(fromBase64(chunk.nonce)),
        additionalData: associatedData(chunk, chunk.nonce),
      },
      await deriveTransferKey(pairingSecret, chunk.transferId),
      toArrayBuffer(chunk.ciphertext),
    ),
  );
  if (plaintext.byteLength !== chunk.plaintextBytes) {
    throw new Error("Decrypted chunk byte count does not match its metadata.");
  }
  return plaintext;
}

export async function sha256Hex(bytes: Uint8Array): Promise<string> {
  const digest = await crypto.subtle.digest("SHA-256", toArrayBuffer(bytes));
  return Buffer.from(digest).toString("hex");
}

function deriveTransferKey(
  pairingSecret: string,
  transferId: string,
): Promise<CryptoKey> {
  const cacheKey = `${pairingSecret}\u0000${transferId}`;
  let key = keyCache.get(cacheKey);
  if (key) return key;
  key = (async () => {
    const baseKey = await crypto.subtle.importKey(
      "raw",
      textEncoder.encode(pairingSecret),
      "PBKDF2",
      false,
      ["deriveKey"],
    );
    return crypto.subtle.deriveKey(
      {
        name: "PBKDF2",
        salt: textEncoder.encode(`${kdfSaltPrefix}\u0000${transferId}`),
        iterations: kdfIterations,
        hash: "SHA-256",
      },
      baseKey,
      { name: "AES-GCM", length: 256 },
      false,
      ["encrypt", "decrypt"],
    );
  })();
  keyCache.set(cacheKey, key);
  return key;
}

function associatedData(
  metadata: TransferChunkMetadata,
  nonce: string,
): ArrayBuffer {
  return toArrayBuffer(
    textEncoder.encode(
      JSON.stringify({
        v: 1,
        transferId: metadata.transferId,
        fileId: metadata.fileId,
        offset: metadata.offset,
        plaintextBytes: metadata.plaintextBytes,
        nonce,
      }),
    ),
  );
}

function toBase64(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("base64");
}

function fromBase64(value: string): Uint8Array {
  return new Uint8Array(Buffer.from(value, "base64"));
}

function toArrayBuffer(bytes: Uint8Array): ArrayBuffer {
  const buffer = new ArrayBuffer(bytes.byteLength);
  new Uint8Array(buffer).set(bytes);
  return buffer;
}

File: app/lib/src/shared/transfer_crypto.dart
import 'dart:convert';
import 'dart:math';

import 'package:cryptography/cryptography.dart';

class TransferChunkMetadata {
  const TransferChunkMetadata({
    required this.transferId,
    required this.fileId,
    required this.offset,
    required this.plaintextBytes,
  });

  final String transferId;
  final String fileId;
  final int offset;
  final int plaintextBytes;
}

class EncryptedTransferChunk extends TransferChunkMetadata {
  const EncryptedTransferChunk({
    required super.transferId,
    required super.fileId,
    required super.offset,
    required super.plaintextBytes,
    required this.nonce,
    required this.ciphertext,
  });

  final String nonce;

  /// AES-GCM ciphertext followed by its 16-byte authentication tag.
  final List<int> ciphertext;
}

class TransferCrypto {
  TransferCrypto({Random? random}) : _random = random ?? Random.secure();

  static final _algorithm = AesGcm.with256bits();
  static final _kdf = Pbkdf2(
    macAlgorithm: Hmac.sha256(),
    iterations: 200000,
    bits: 256,
  );
  static const _saltPrefix = 'vidyut-v1-file-transfer';

  final Random _random;
  final _derivedKeys = <String, Future<SecretKey>>{};

  Future<EncryptedTransferChunk> encrypt({
    required TransferChunkMetadata metadata,
    required List<int> plaintext,
    required String pairingSecret,
  }) async {
    if (plaintext.length != metadata.plaintextBytes) {
      throw const FormatException(
        'Chunk byte count does not match its metadata.',
      );
    }
    final nonce = _randomBytes(12);
    final nonceBase64 = base64Encode(nonce);
    final box = await _algorithm.encrypt(
      plaintext,
      secretKey: await _deriveKey(pairingSecret, metadata.transferId),
      nonce: nonce,
      aad: _associatedData(metadata, nonceBase64),
    );
    return EncryptedTransferChunk(
      transferId: metadata.transferId,
      fileId: metadata.fileId,
      offset: metadata.offset,
      plaintextBytes: metadata.plaintextBytes,
      nonce: nonceBase64,
      ciphertext: [...box.cipherText, ...box.mac.bytes],
    );
  }

  Future<List<int>> decrypt({
    required EncryptedTransferChunk chunk,
    required String pairingSecret,
  }) async {
    if (chunk.ciphertext.length < 16) {
      throw const FormatException('Transfer ciphertext is too short.');
    }
    final macOffset = chunk.ciphertext.length - 16;
    final plaintext = await _algorithm.decrypt(
      SecretBox(
        chunk.ciphertext.sublist(0, macOffset),
        nonce: base64Decode(chunk.nonce),
        mac: Mac(chunk.ciphertext.sublist(macOffset)),
      ),
      secretKey: await _deriveKey(pairingSecret, chunk.transferId),
      aad: _associatedData(chunk, chunk.nonce),
    );
    if (plaintext.length != chunk.plaintextBytes) {
      throw const FormatException(
        'Decrypted chunk byte count does not match its metadata.',
      );
    }
    return plaintext;
  }

  Future<String> sha256Hex(List<int> bytes) async {
    final hash = await Sha256().hash(bytes);
    return hash.bytes
        .map((byte) => byte.toRadixString(16).padLeft(2, '0'))
        .join();
  }

  Future<SecretKey> _deriveKey(String pairingSecret, String transferId) {
    final cacheKey = '$pairingSecret\u0000$transferId';
    return _derivedKeys.putIfAbsent(cacheKey, () {
      return _kdf.deriveKey(
        secretKey: SecretKey(utf8.encode(pairingSecret)),
        nonce: utf8.encode('$_saltPrefix\u0000$transferId'),
      );
    });
  }

  List<int> _associatedData(TransferChunkMetadata metadata, String nonce) {
    return utf8.encode(
      jsonEncode({
        'v': 1,
        'transferId': metadata.transferId,
        'fileId': metadata.fileId,
        'offset': metadata.offset,
        'plaintextBytes': metadata.plaintextBytes,
        'nonce': nonce,
      }),
    );
  }

  List<int> _randomBytes(int length) {
    return List<int>.generate(length, (_) => _random.nextInt(256));
  }
}
